operator/v1/types_machineconfiguration.go

```diff
@@ -120,6 +130,39 @@ type MachineConfigurationList struct {
 	Items []MachineConfiguration `json:"items"`
 }
 
+// IrreconcilableMachineConfigValidationOverridesStorage defines available storage irreconcilable overrides.
+// +kubebuilder:validation:Enum=Disks;FileSystems;Raid
+type IrreconcilableMachineConfigValidationOverridesStorage string
+
+const (
+	// Disks enables changes to the `spec.config.storage.disks` section of MachineConfig CRs.
+	IrreconcilableMachineConfigValidationOverridesStorageDisks IrreconcilableMachineConfigValidationOverridesStorage = "Disks"
+
+	// FileSystems enables changes to the `spec.config.storage.filesystems` section of MachineConfig CRs.
+	IrreconcilableMachineConfigValidationOverridesStorageFileSystems IrreconcilableMachineConfigValidationOverridesStorage = "FileSystems"
+
+	// Raid enables changes to the `spec.config.storage.raid` section of MachineConfig CRs.
+	IrreconcilableMachineConfigValidationOverridesStorageRaid IrreconcilableMachineConfigValidationOverridesStorage = "Raid"
+)
+
+// IrreconcilableMachineConfigValidationOverrides holds the irreconcilable validations overrides to be applied on
+// each rendered MachineConfig generation.
+type IrreconcilableMachineConfigValidationOverrides struct {
+	// storage can be used to allow making irreconcilable changes to the selected sections under the
+	// `spec.config.storage` field of MachineConfig CRs
+	// It must have at least one item, may not exceed 3 items and must not contain duplicates.
+	// Allowed element values are "Disks", "FileSystems", "Raid" and omitted.
+	// When contains "Disks" changes to the `spec.config.storage.disks` section of MachineConfig CRs are allowed.
+	// When contains "FileSystems" changes to the `spec.config.storage.filesystems` section of MachineConfig CRs are allowed.
+	// When contains "Raid" changes to the `spec.config.storage.raid` section of MachineConfig CRs are allowed.
+	// When omitted changes to the `spec.config.storage` section are forbidden.
+	// +optional
+	// +listType=set
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=3
+	Storage []IrreconcilableMachineConfigValidationOverridesStorage `json:"storage"`
```

Comment:
Because the empty value `[]` is not valid, we want `omitempty` on the field here so we never marshal an empty value that structured clients would use.

Reply:
That being said, if you are going for discoverability we would probably want to allow the empty value now that I think about it.

IIRC, we are looking to move away from discoverability though. @JoelSpeed am I recalling correctly?
It looks like this API has historically followed the discoverability guidance - should we remain consistent here?